File: algorithms_and_data_structures/seminar-2/heapsort.java
public class heapsort {
    public void sort(int array[]) {
        int n = array.length;
        for (int i = n / 2 - 1; i >= 0; i--)
            heapify(array, n, i);
        for (int i = n - 1; i >= 0; i--) {
            int temp = array[0];
            array[0] = array[i];
            array[i] = temp;

            heapify(array, i, 0);
        }
    }

    void heapify(int array[], int n, int i) {
        int largest = i;
        int l = 2 * i + 1;
        int r = 2 * i + 2;

        if (l < n && array[l] > array[largest])
            largest = l;

        if (r < n && array[r] > array[largest])
            largest = r;
        if (largest != i) {
            int swap = array[i];
            array[i] = array[largest];
            array[largest] = swap;

            heapify(array, n, largest);
        }
    }

    // Управляющая программа
    public static void main(String args[]) {
        int[] array = new int[] {
            4, 2, 5, 8, 1, 9, 2, 3, 6, 8, 5
    };
        int n = array.length;

        heapsort ob = new heapsort();
        ob.sort(array);

        for (int i = 0; i < n; ++i)
            System.out.print(array[i] + " ");
    }
}

File: algorithms_and_data_structures/lecture-2/heapsort.py
# Сортировка кучей
# Сложность алгоритма = O(n * log n)

# Реализация пирамидальной сортировки на Python

# Процедура для преобразования в двоичную кучу поддерева с корневым узлом i, что является индексом в arr[]. n - размер кучи
def heapify(arr, n, i):
    largest = i  # Initialize largest as root
    l = 2 * i + 1   # left = 2*i + 1
    r = 2 * i + 2   # right = 2*i + 2

# Проверяем существует ли левый дочерний элемент больший, чем корень

    if l < n and arr[i] < arr[l]:
        largest = l

    # Проверяем существует ли правый дочерний элемент больший, чем корень

    if r < n and arr[largest] < arr[r]:
        largest = r

    # Заменяем корень, если нужно
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]  # свап

        # Применяем heapify к корню.
        heapify(arr, n, largest)

# Основная функция для сортировки массива заданного размера


def heapSort(arr):
    n = len(arr)

    # Построение max-heap.
    for i in range(n, -1, -1):
        heapify(arr, n, i)

    # Один за другим извлекаем элементы
    for i in range(n-1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]  # свап
        heapify(arr, i, 0)


# Управляющий код для тестирования
arr = [4, 2, 5, 8, 1, 9, 2, 3, 6, 8, 5]
heapSort(arr)
n = len(arr)
print(f"отсортированный массив: {arr}")
